src/utils/OnboardingProvider.tsx

```diff
@@ -0,0 +1,293 @@
+    const setLanguages = async (newLangs: ProfileLanguage[]) => {
+      if (profile.user_id === blankProfile.user_id)
+        throw new Error(
+          `Expected user_id to be valid but got: ${profile.user_id}`,
+        );
+
+      const toDelete = newLangs.filter(
+        l1 => !languages.find(l2 => l1.iso_code === l2.iso_code),
+      );
+
+      await Promise.all([
+        deleteLanguages(profile.user_id, toDelete),
+        upsertLanguages(languages).then(data => setProfileLangs(data)),
+      ]);
+    };
+
+    /**
+     * Takes in a list of ProfileRole
+     * and replaces roles with it entirely,
+     * simultaneously updating the database.
+     * Requires RLS delete, update, insert on profiles-roles.
+     */
+    const setRoles = async (newRoles: ProfileRole[]) => {
+      if (profile.user_id === blankProfile.user_id)
+        throw new Error(
+          `Expected user_id to be valid but got: ${profile.user_id}`,
+        );
+
+      const toDelete = newRoles.filter(
+        r1 => !roles.find(r2 => r1.role === r2.role),
+      );
+
+      await Promise.all([
+        deleteRoles(profile.user_id, toDelete),
+        upsertRoles(roles).then(data => setProfileRoles(data)),
+      ]);
+    };
+
+    const val: OnboardingContextType = {
+      progress,
+      profile,
+      languages,
+      roles,
+      flow,
+      canContinue,
+      setFlow,
+      setProgress,
+      updateProfile,
+      flushProfileData,
+      setLanguages,
+      setRoles,
+      setCanContinue,
+    };
+
+    return val;
+  }, [progress, profile, flow, languages, roles, canContinue]);
+
+  return (
+    <OnboardingContext.Provider value={providerValue}>
+      {children}
+    </OnboardingContext.Provider>
+  );
+}
+
+/**
+ * EXAMPLE USAGE:
+ */
+
+/**
+ * app/test/layout.tsx
+
+'use client';
+
+import { ReactNode } from 'react';
+import OnboardingProvider from '@/utils/OnboardingProvider';
+
+export default function TestLayout({ children }: { children: ReactNode }) {
+  return <OnboardingProvider>{children}</OnboardingProvider>;
+}
+
+*/
+
+/**
+ * app/test/page.tsx
+
+'use client';
+
+import { useContext, useState } from 'react';
+import { OnboardingContext } from '@/utils/OnboardingProvider';
+
+export default function Page() {
+  const onboarding = useContext(OnboardingContext);
+  const [firstName, setFirstName] = useState('');
+  const [lastName, setLastName] = useState('');
+  const [hoursPerMonth, setHoursPerMonth] = useState(0);
+  const [location, setLocation] = useState('');
+  const [startDate, setStartDate] = useState('');
+  const [error, setError] = useState<string>('');
+
+  const updateProfile = async () => {
+    if (!onboarding) {
+      setError('No onboarding context detected');
+      return;
+    }
+    setError('');
+
+    await onboarding.updateProfile({
+      first_name: firstName,
+      last_name: lastName,
+      hours_per_month: hoursPerMonth,
+      location,
+      start_date: startDate,
+    });
+
+    setTimeout(() => {
```

Comment:
could you explain what this is doing? are we flushing the profile on supabase in the case that there's an error in the onboarding flow?

Reply:
This is for testing only, but the flushing data is meant to update Supabase when we want to. The idea is to only push to Supabase at the end and update states only.